components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/impersonation/validators/ResidentOrganizationValidator.java

```diff
@@ -0,0 +1,114 @@
+
+import static org.wso2.carbon.identity.application.common.util.IdentityApplicationConstants.Error.INVALID_REQUEST;
+
+/**
+ * Validator to validate impersonation requests for users from different resident organizations.
+ * This validator checks if the impersonated user belongs to a different organization and updates the
+ * impersonated user details accordingly.
+ */
+public class ResidentOrganizationValidator implements ImpersonationValidator {
+
+    private static final String NAME = "ResidentOrganizationValidator";
+    private static final Log LOG = LogFactory.getLog(ResidentOrganizationValidator.class);
+
+    @Override
+    public int getPriority() {
+
+        return 300;
+    }
+
+    @Override
+    public String getImpersonationValidatorName() {
+
+        return NAME;
+    }
+
+    @Override
+    public ImpersonationContext validateImpersonation(ImpersonationContext impersonationContext)
+            throws IdentityOAuth2Exception {
+
+        try {
+            AuthenticatedUser impersonatingActor = impersonationContext.getImpersonationRequestDTO().getImpersonator();
+            ImpersonatedUser impersonatedUser = impersonatingActor.getImpersonatedUser();
+            String impersonatedUserResidentOrg = impersonatingActor.getUserResidentOrganization();
+            String impersonatedUserId = impersonatedUser.getUserId();
+
+            if (impersonatedUserResidentOrg != null) {
+                UserAssociation association = getUserAssociation(impersonatedUserId, impersonatedUserResidentOrg);
+                if (association != null) {
+                    // User from a different org.
+                    impersonatedUserResidentOrg = association.getUserResidentOrganizationId();
+                    impersonatedUserId = association.getAssociatedUserId();
+                    try {
+                        if (OrganizationManagementUtil.isOrganization(impersonatedUserResidentOrg)) {
+                            // Org is another sub org.
+                            impersonatedUser.setUserResidentOrganization(impersonatedUserResidentOrg);
+                        }
+                    } catch (OrganizationManagementException e) {
+                        throw new IdentityOAuth2ClientException(INVALID_REQUEST.getCode(),
+                                "Invalid User Id provided for the request. Unable to find the user for given " +
+                                        "user id : " + impersonatedUserId + " organization : "
+                                        + impersonatedUserResidentOrg, e);
+                    }
+                } else {
+                    // User from the same sub org.
+                    impersonatedUser.setUserResidentOrganization(impersonatedUserResidentOrg);
+                }
+                // User from parent org.
+                impersonatedUser.setUserId(impersonatedUserId);
+                impersonatedUser.setUserResidentOrganization(null);
```

Comment:
Won't impersonatedUser.setUserResidentOrganization(...) always be null here, since it's being set to null unconditionally at the end of the block, overriding any earlier assignments?